### PR TITLE
Introspection 1/7, Support for signature in registered procedures

### PR DESCRIPTION
**Pull request by bombela**
The router will hold onto extra informations about handlers:
 - encoding of the handler
 - signature of the handler

The yarpc thriftrw plugins will also register a signature automatically.

The signature is a simple string, hopefully you write in there some Go looking
function definition. This is convenient when you use thrift (or any IDL system
doing the generation for you) because most people wont likely bother adding a
signature to a json handler by hand.

Which means, we could use reflection maybe. I tried keeping things simple for now.

### CODE REVIEW

File: encoding/thrift/register.go [hunk 1]
```diff
@@ -103,10 +107,13 @@ func BuildProcedures(s Service, opts ...RegisterOption) []transport.Procedure {
 			Protocol:     proto,
 			Enveloping:   rc.Enveloping,
 		})
+		signature, _ := s.Signatures[methodName]
```
Comment (akshayjshah) [inline encoding/thrift/register.go]: No need for the blank identifier; this is the same as `signature := s.Signatures[methodName]`, so you might as well do it inline in the procedure literal below.

File: encoding/thrift/register.go [hunk 2]
```diff
@@ -79,6 +79,10 @@ type Service struct {
 	Name          string
 	Methods       map[string]UnaryHandler
 	OnewayMethods map[string]OnewayHandler
+
+	// Snippet of Go code representing the function definition of the different
+	// handlers. This is useful for introspection.
+	Signatures map[string]string
```
Comment (akshayjshah) [inline encoding/thrift/register.go]: To me, this is an odd place to stash this information. I'd expect each `thrift.UnaryHandler` and `thrift.OnewayHandler` to implement `Signature() string`.

File: api/transport/router.go [hunk 3]
```diff
@@ -60,3 +67,22 @@ type RouteTable interface {
 	// Registers zero or more procedures with the route table.
 	Register([]Procedure)
 }
+
+// ProceduresByServiceProcedure offers to sort a slice of Procedure by Service
+// & Procedure.
+type ProceduresByServiceProcedure []Procedure
```
Comment (bombela) [inline api/transport/router.go]: I reused the existing naming + some changes that got intertwined during a rebase changing the naming to this as well. I am not against a better name!

File: encoding/thrift/register.go (hunk 2 above)
Comment (bombela) [inline encoding/thrift/register.go]: We use thrift.*Handler(s) as function/method wrappers. Adding a Signature() method would then require to build some structure during code-gen to hold the pair (function, signature) which is a bigger change to code-gen. But that's arguable, @abhinav what are your thoughts?

File: encoding/thrift/register.go (hunk 1 above)
Comment (bombela) [inline encoding/thrift/register.go]: Without the blank identifier, it will panic instead of returning an empty string if the entry cannot be found.

File: encoding/thrift/register.go (hunk 2 above)
Comment (bombela) [inline encoding/thrift/register.go]: After discussion with @abhinav , conclusion is: let's refactor this to use a single map of procname to struct {} isntead.

File: encoding/thrift/register.go (hunk 1 above)
Comment (bombela) [inline encoding/thrift/register.go]: my bad it wont, it will return an empty string. I dono why I thought that.

File: router.go [hunk 4]
```diff
@@ -64,22 +64,18 @@ func (m MapRouter) Register(rs []transport.Procedure) {
...
-		procs = append(procs, transport.Procedure{
-			Service:     sp.service,
-			Name:        sp.procedure,
-			HandlerSpec: handler,
-		})
+	for _, v := range m.entries {
+		procs = append(procs, v)
 	}
-	sort.Sort(byServiceProcedure(procs))
+	sort.Sort(transport.ProceduresByServiceProcedure(procs))
```
Comment (akshayjshah) [inline router.go]: instead,

```
sort.Slice(procs, func(i, j int) { procs[i].Less(procs[j]) })
```

File: api/transport/router.go [hunk 5]
```diff
@@ -60,3 +67,22 @@ type RouteTable interface {
...
+func (sp ProceduresByServiceProcedure) Less(i int, j int) bool {
+	if sp[i].Service == sp[j].Service {
+		return sp[i].Name < sp[j].Name
+	}
+	return sp[i].Service < sp[j].Service
+}
+
+func (sp ProceduresByServiceProcedure) Swap(i int, j int) {
+	sp[i], sp[j] = sp[j], sp[i]
+}
```
Comment (bombela) [inline api/transport/router.go]: done

File: encoding/thrift/register.go [hunk 6]
```diff
@@ -39,45 +40,38 @@ func Register(r transport.RouteTable, rs []transport.Procedure) {
...
 
-// UnaryHandler represents a Thrift request handler. It speaks in raw Thrift payloads.
-//
-// Users should use the server package generated by the code generator rather
-// than using this directly.
-type UnaryHandler interface {
-	Handle(ctx context.Context, body wire.Value) (Response, error)
-}
+// UnaryHandler is a convenience type alias for functions that act as Handlers.
+type UnaryHandler func(context.Context, wire.Value) (Response, error)
```
Comment (abhinav) [inline encoding/thrift/register.go]: Please update the doc. This is not a convenience type alias anymore. It is what UnaryHandler was before.

File: encoding/thrift/register.go [hunk 7]
```diff
@@ -39,45 +40,38 @@ func Register(r transport.RouteTable, rs []transport.Procedure) {
...
 
-// OnewayHandler represents a Thrift request handler. It speaks in raw Thrift payloads.
-//
-// Users should use the server package generated by the code generator rather
-// than using this directly.
-type OnewayHandler interface {
-	HandleOneway(ctx context.Context, body wire.Value) error
-}
+// OnewayHandler is a convenience type alias for functions that act as OnewayHandlers.
+type OnewayHandler func(context.Context, wire.Value) error
```
Comment (abhinav) [inline encoding/thrift/register.go]: Same

File: encoding/thrift/inbound_test.go [hunk 8]
```diff
@@ -130,16 +129,18 @@ func TestThriftHandler(t *testing.T) {
...
+			error) {
 
-		if tt.expectHandle {
-			handler.EXPECT().Handle(gomock.Any(), requestBody).
-				Return(Response{
-					Body:               fakeEnveloper(tt.responseEnvelopeType),
-					IsApplicationError: tt.responseIsAppError,
-				}, nil)
+			if tt.expectHandle {
+				assert.Equal(t, requestBody, w)
```
Comment (abhinav) [inline encoding/thrift/inbound_test.go]: Not what this was testing. This was testing that if `tt.expectHandle` is true,
then this function is called with these arguments. Otherwise, it fails if this
function is called.

You should just make the function check the requestBody, etc. and track
whether the function was called in a boolean outside. Then just match that
bool against `tt.expectHandle`.

File: api/transport/router.go [hunk 9]
```diff
@@ -37,6 +37,21 @@ type Procedure struct {
 
 	// HandlerSpec specifiying which handler and rpc type.
 	HandlerSpec HandlerSpec
+
+	// Encoding of the handler, for introspection.
+	Encoding Encoding
+
+	// Signature of the handler, for introspection. This should be a snippet of
+	// Go code representing the function definition.
```
Comment (breerly) [inline api/transport/router.go]: Not sure what you mean "snippet of Go code" in this context - don't you mean "a string"?

File: api/transport/router.go [hunk 10]
```diff
@@ -37,6 +37,21 @@ type Procedure struct {
...
+	Signature string
+}
+
+// Less returns true if a.(Service, Name) < b.(Service, Name).
+func (a Procedure) Less(b Procedure) bool {
```
Comment (breerly) [inline api/transport/router.go]: Would be useful to described *why* this function needs to exist, or *how* it's used.

File: encoding/thrift/register.go [hunk 11]
```diff
@@ -39,45 +40,38 @@ func Register(r transport.RouteTable, rs []transport.Procedure) {
...
 
-// UnaryHandlerFunc is a convenience type alias for functions that act as Handlers.
-type UnaryHandlerFunc func(context.Context, wire.Value) (Response, error)
+// HandlerSpec represents the handler behind a Thrift service method.
+type HandlerSpec struct {
```
Comment (breerly) [inline encoding/thrift/register.go]: Is this related to this PR? Did you rebase OK?
